### PR TITLE
Wemos Lolin C3 Mini Board

### PR DESCRIPTION
**Pull request by askpatrickw**
This is board definition for the Lolin C3 Mini

TODO:
(Reworking all this based on a different baseline board defition (QtPyC3)

- [x] Should this be "lolin" or "wemos"? -- Using Wemos for company and using lolin in board name.
- [x] Question: I see [IGNORE_PIN_ in the docs](https://learn.adafruit.com/how-to-add-a-new-board-to-circuitpython/customizing-the-board-files), but not in any Espressif board definitions. Should we ignore that for Espressif boards? -- 
- [x] Question: Adafruit_CircuitPython_NeoPixel is frozen, should it be? -- No. This board does not have a Neopixel
- [x] VID and PID are currently the same values as the the Lolin S2 Mini. Not sure if that is correct. NOPE, need a new PID, requested from Espressif.
- [x] QUESTION is the C3 Bin only? I set the build to BIN only as other C3 boards appear to be and the CI Failed. Yes they are [BIN only](https://github.com/adafruit/circuitpython/commit/ce3f1dda16300b92c53aa02d04f6a5321276aec6).

### Wemos Lolin C3 Mini Testing

- [x] SafeMode Boot by tapping Boot Mode on power on
- [x] Pin 0: A0: 
- [x] Pin 1: A1: 
- [x] Pin 2: A2:  Pin 2 is always pulled high according to the schematic
- [x] Pin 3: A3: 
- [x] Pin 4: A4: 
- [x] Pin 5: A5: SDA: i2c.scan() worked
- [x] Pin 6: SCL: i2c.scan() worked
- [x] Pin 7: board.LED on\off works
- [x] Pin 8: Pin 8 is always pulled high according to the schematic
- [x] Pin 10: 
- [x] UART (PIN: 20,21/A6,A7)
- [x] WIFI
	- [x] Scan Networks
	- [x] Connect: FAILING ConnectionError: Unknown failure 205 -- I don't think has to do with this board. Fixed by #6540 
- [ ] BLE
- [ ] rshell/ampy access: FAILING, they do not connect

----
image

**Snippet of the Schematic**

Some helper docs:
* [C3 Mini Homepage](https://www.wemos.cc/en/latest/c3/c3_mini.html)
* [C3 Mini Schematic](https://www.wemos.cc/en/latest/_static/files/sch_c3_mini_v1.0.0.pdf)
* [C3 Datasheet](https://www.espressif.com/sites/default/files/documentation/esp32-c3_datasheet_en.pdf)
* [C3 Mini MicroPython Download Page](https://micropython.org/download/LOLIN_C3_MINI/)
* [C3 MicroPython Port](https://github.com/micropython/micropython/tree/master/ports/esp32/boards/LOLIN_C3_MINI)

### CODE REVIEW

**Comment by kylefmohr** [comment]
Hi, looking forward to testing this out later this afternoon! Regarding question #2, I would think it would be 'Lolin', based on their ESP32-S2 board which is supported https://circuitpython.org/board/lolin_s2_mini/

**Comment by askpatrickw** [comment]
Yeah the Company appears to be called Wemos and the RTD url is Wemos.
https://www.wemos.cc/en/latest/c3/c3_mini.html

The Aliepxress store is Lolin.
https://lolin.aliexpress.com/store/1331105 and the descriptions mention Wemos.

But sometimes the boards have Lolin in the name on the Wemos RTD site, and in the silkscreen, sometimes they don't. https://www.wemos.cc/en/latest/index.html

🤷‍♂️ -- I'm going to leave it as Lolin unless someone corrects it.

**Comment by askpatrickw** [comment]
I'm goin to park this PR as a draft unless someone has an idea on how to get a CPY file onto the board without BLE workflow. 

If you'd rather I close it and re-open later, that's OK too.

**Comment by askpatrickw** [comment]
NVM about being blocked...
Neradoc pointed me at the QTPYC3 to compare to, and now I have REPL access so I'm not blocked.

File: ports/espressif/boards/lolin_c3_mini/mpconfigboard.mk [hunk 1]
```diff
@@ -0,0 +1,12 @@
+USB_VID = 0x303A
+USB_PID = 0x80F5
+USB_PRODUCT = "Lolin C3 Mini"
+USB_MANUFACTURER = "Wemos"
```
Comment (prplz) [inline ports/espressif/boards/lolin_c3_mini/mpconfigboard.mk:4]: C3 does not use these as it does not have native USB.
For bluetooth, C3 boards require a creator/creation ID: https://github.com/creationid/

Comment (askpatrickw) [reply ports/espressif/boards/lolin_c3_mini/mpconfigboard.mk:4]: Thanks @prplz !

For future me, when I get to this: Scott talks about Creation ID on the Deep Dive

https://github.com/adafruit/deep-dive-notes/search?q=creation+id

**Comment by askpatrickw** [comment]
Pin 2 an 8 are always pulled high according to the schematic so I'm checking those off the test list since I see that this is true when testing.

**Comment by askpatrickw** [comment]
I'm flipping this to ready for review. It might not be perfect, but I think its working well. The one open item is the Creations ID and Repo, but I think when Scott is back that will be cleaned up and be implemented as I suggested and have it in the files.

File: ports/espressif/boards/lolin_c3_mini/mpconfigboard.mk (hunk 1 above)
Comment (tannewt) [reply ports/espressif/boards/lolin_c3_mini/mpconfigboard.mk:4]: Please remove these defines because they aren't used for C3.

**Review by tannewt — CHANGES_REQUESTED** [review]
One request. Wemos creations repo should be ready for a PR too.

Thanks for adding this board!

**Comment by prplz** [comment]
Your pin naming isn't right. Circuitpython uses whatever the silkscreen says, in this case matching the IO# numbering. So IO0-IO10 should be D0-D10, etc.

**Comment by askpatrickw** [comment]
Thanks @tannewt and @prplz I will look at this next week.

File: ports/espressif/boards/lolin_c3_mini/mpconfigboard.mk (hunk 1 above)
Comment (askpatrickw) [reply ports/espressif/boards/lolin_c3_mini/mpconfigboard.mk]: This is fixed in this commit:
https://github.com/adafruit/circuitpython/pull/6256/commits/47c5b3da1f4e8d196472bb0635ecbe2e72508eef

**Comment by askpatrickw** [comment]
> 

@prplz  I'm back from my trip and hoping to finish this up ASAP. Now the I'm looking at the board, I don't see any "D" labels on the silscreen, so I'm confused. If you want to "@" me on Discord, we can do that too. Thanks for your help!

File: ports/espressif/boards/lolin_c3_mini/mpconfigboard.mk [hunk 2]
```diff
@@ -1,7 +1,5 @@
-USB_VID = 0x303A
-USB_PID = 0x80F5
-USB_PRODUCT = "Lolin C3 Mini"
-USB_MANUFACTURER = "Wemos"
+CIRCUITPY_CREATOR_ID =  0x1988_1988
```
Comment (kylefmohr) [inline ports/espressif/boards/lolin_c3_mini/mpconfigboard.mk:1]: It appears that this is causing the build to fail, as it is expecting the CIRCUITPY_CREATOR_ID to be an integer. [Based on another example of somebody using a creator ID](https://github.com/adafruit/circuitpython/blob/14b734be6ed7d8a2644610d0fd899f516deae51d/ports/espressif/boards/lilygo_ttgo_t-01c3/mpconfigboard.mk#L1), it seems you could fix this problem simply by removing the underscore

Comment (askpatrickw) [reply ports/espressif/boards/lolin_c3_mini/mpconfigboard.mk]: Thanks @kylefmohr  !

**Comment by prplz** [comment]
> > 
> 
> @prplz I'm back from my trip and hoping to finish this up ASAP. Now the I'm looking at the board, I don't see any "D" labels on the silscreen, so I'm confused. If you want to "@" me on Discord, we can do that too. Thanks for your help!

Right, on the silkscreen they are just numbers. The naming in pins.c should match the silk, for example you have `{ MP_ROM_QSTR(MP_QSTR_D8), MP_ROM_PTR(&pin_GPIO0) },` but on the silk it's named 0. I said D but the convention seems to be to use "IO" for esp32 boards.
```diff
     { MP_ROM_QSTR(MP_QSTR_A0), MP_ROM_PTR(&pin_GPIO0) },
-    { MP_ROM_QSTR(MP_QSTR_D8), MP_ROM_PTR(&pin_GPIO0) },
+    { MP_ROM_QSTR(MP_QSTR_IO0), MP_ROM_PTR(&pin_GPIO0) },

     { MP_ROM_QSTR(MP_QSTR_A1), MP_ROM_PTR(&pin_GPIO1) },
-    { MP_ROM_QSTR(MP_QSTR_D9), MP_ROM_PTR(&pin_GPIO1) },
+    { MP_ROM_QSTR(MP_QSTR_IO1), MP_ROM_PTR(&pin_GPIO1) },
```
and so on.

**Comment by evildave666** [comment]
> > > 
> > 
> > 
> > @prplz I'm back from my trip and hoping to finish this up ASAP. Now the I'm looking at the board, I don't see any "D" labels on the silscreen, so I'm confused. If you want to "@" me on Discord, we can do that too. Thanks for your help!
> 
> Right, on the silkscreen they are just numbers. The naming in pins.c should match the silk, for example you have `{ MP_ROM_QSTR(MP_QSTR_D8), MP_ROM_PTR(&pin_GPIO0) },` but on the silk it's named 0. I said D but the convention seems to be to use "IO" for esp32 boards.
> 

There might be an argument for retaining an alternate "D" pin notation to match that of the D1 mini board, as there's an existing ecosystem of shields with markings matching those on the D1 board.

**Comment by prplz** [comment]
> There might be an argument for retaining an alternate "D" pin notation to match that of the D1 mini board, as there's an existing ecosystem of shields with markings matching those on the D1 board.

 I see, maybe keep the D names and add IO as well?

**Comment by askpatrickw** [comment]
OK, I wasn't understanding this 100%, but I took a picture of the board, zoomed in, followed the traces and I get it now. In the passing of time, I have no idea why I used those D Mappings. Latest commit maps the silkscreen numbers to the correct pins (D to GPIO).

The IO:GPIO mappings are 1:1 aligned. There are always the same according to my reading of the schematic.

With two possible extraneous extra IOs. IO20, IO21 are not on the schematic for  GPIO20, GPIO21, they are labeled RX and TX instead. Let me know if those IO mappings should be removed or if they are also valid.

thank so much!

**Review by kylefmohr — COMMENTED** [review]
I think the build is failing because your fork is a ways behind the main branch. If you fetch the upstream, it may fix the build issues.

**Comment by kylefmohr** [comment]
> I think the build is failing because your fork is a ways behind the main branch. If you fetch the upstream, it may fix the build issues.

I was able to confirm this by forking [adafruit:main](https://github.com/adafruit/circuitpython) with the most recent commits and applying the changes made by [askpatrickw:lolin-c3-mini](https://github.com/askpatrickw/circuitpython/tree/lolin-c3-mini). Everything built successfully [here](https://github.com/kylefmohr/circuitpython/actions/runs/2508347296)

**Comment by evildave666** [comment]
This board seems to have a hardware issue that it needs to have TX power explicitly adjusted away from default for wifi to function (I've verified this in Micropython - won't connect at default power but after adjusting it works fine). I don't believe the circuitpython wifi library for ESP32 has a runtime setting for this so it might have to be set in board.c

**Comment by kylefmohr** [comment]
> This board seems to have a hardware issue that it needs to have TX power explicitly adjusted away from default for wifi to function (I've verified this in Micropython - won't connect at default power but after adjusting it works fine and a similar caveat exists in Wemos' arduino documentation). I don't believe the circuitpython wifi library for ESP32 has a runtime setting for this (Micropython has sta_if.config(txpower=xxx) but an equivalent does not appear to exist in Circuitpython) so it might have to be set in board.c

Interesting, that explains why my build wouldn't connect to Wi-Fi. [Here's a link to their documentation regarding this for MicroPython](https://www.wemos.cc/en/latest/tutorials/c3/get_started_with_micropython_c3.html#wifi)

**Comment by askpatrickw** [comment]
> I think the build is failing because your fork is a ways behind the main branch. If you fetch the upstream, it may fix the build issues.

Thanks @kylefmohr  this PR got old... thanks for the reminder.

**Comment by askpatrickw** [comment]
@evildave666  and @kylefmohr  should we split that WIFI item out into a new issue?
Frequently boards are made live without "all the things" working.

**Comment by evildave666** [comment]
> @evildave666 and @kylefmohr should we split that WIFI item out into a new issue? Frequently boards are made live without "all the things" working.

I had a short discussion today in discord with todbot and neradoc about how to possibly address it. It might be as simple as setting an esp-idf variable in the sdkconfig file in the board directory. CONFIG_ESP32_PHY_MAX_WIFI_TX_POWER might be the appropriate one to set.

**Comment by tannewt** [comment]
Please try the tx power suggestion from @evildave666 and then I'll review. I suspect it will work.

**Comment by askpatrickw** [comment]
Back at my desk with access to my board. No, that fix didn't work.
I'm capturing some notes about this as I look into it:

- CONFIG_ESP32_PHY_MAX_WIFI_TX_POWER changed to [CONFIG_ESP_PHY_MAX_WIFI_TX_POWER](https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32/api-reference/kconfig.html?highlight=config_esp_phy_max_wifi_tx_power#config-esp-phy-max-wifi-tx-power) and the valid values are 10-20, not 8.5 as mentioned in MicroPython.

- MicroPython [appears to do a calculation](https://github.com/micropython/micropython/blob/ccaf1978072a1d8e63b5e2ee350f15d875dfc1fd/ports/esp32/network_wlan.c#L473) using [the 8.5 they recommend](https://www.wemos.cc/en/latest/tutorials/c3/get_started_with_micropython_c3.html#wifi) (resulting in the value 34) and then call [esp_wifi_set_max_tx_power](https://docs.espressif.com/projects/esp-idf/en/v4.4/esp32/api-reference/network/esp_wifi.html?highlight=esp_wifi_set_max_tx_power#_CPPv425esp_wifi_set_max_tx_power6int8_t)
- `Mapping Table {Power, max_tx_power} = {{8, 2}, {20, 5}, {28, 7}, {34, 8}, {44, 11}, {52, 13}, {56, 14}, {60, 15}, {66, 16}, {72, 18}, {80, 20}}.` is then used with the calculated value (34), so the power is set to 8, I think... 

I'm not understanding how the 10-20 range in the docs CONFIG_ESP_PHY_MAX_WIFI_TX_POWER is equivalent to the calculations and resulting settings in the way MP sets this.

**Comment by kylefmohr** [comment]
@askpatrickw those are good notes, thanks for including links to the supporting documentation as well. The way Espressif handles this is convoluted, to say the least. 

I haven't done the research or testing to verify this, but is it possible the `CONFIG_ESP32_PHY_MAX_WIFI_TX_POWER` variable doesn't come into play until you try to set a value over the limit? Put another way, is that variable only for setting the maximum wi-fi transmit power? 

It looks like the MP workaround sets the actual transmit power at runtime. Ideally, we would pre-set the actual transmit power as part of this PR so that the end-user doesn't have to worry about it, or, barring that, figure out a way to set it at runtime just like MP does, and then clearly document that procedure. Unfortunately I'm currently unsure how to do either of those things. Hopefully I will have some time tonight to dig into this some more and help figure out a solution.

**Comment by askpatrickw** [comment]
@kylefmohr Would the correct approach be to add a call to [esp_wifi_set_max_tx_power](https://github.com/espressif/esp-idf/blob/master/components/esp_wifi/include/esp_wifi.h#L913) in [wifi.radio.connect](https://github.com/adafruit/circuitpython/blob/main/ports/espressif/common-hal/wifi/Radio.c#L235)?

I'm very much not sure I'm linking to wifi.radio.connect() there, but I see other calls to esp_wifi in that area.

If wifi.radio.connect() is common across all ports, I'm not sure of the right way to add something port specific.
If there is only one value that works, I could see that going in the board config, but if it is meant to be tunable, I don't know the correct approach for that.

**Comment by evildave666** [comment]
> I haven't done the research or testing to verify this, but is it possible the `CONFIG_ESP32_PHY_MAX_WIFI_TX_POWER` variable doesn't come into play until you try to set a value over the limit? Put another way, is that variable only for setting the maximum wi-fi transmit power?

According to todbot it sets the default as well. See [discord discussion](https://discord.com/channels/327254708534116352/537365702651150357/989336569901109328)

**Comment by askpatrickw** [comment]
Thank you @todbot! I'll try that later tonight or tomorrow night.

**Comment by tannewt** [comment]
> ooof I'm looking in: [esp-idf/componenents/esp_phy/esp32s2/include/phy_init_data.h](https://github.com/espressif/esp-idf/blob/944c01eef4fbba693f991f9d033cda3f59ca82c9/components/esp_phy/esp32c3/include/phy_init_data.h) and similar files grepping for "TX_POWER" and am really wishing CircuitPython had access to `esp_wifi_set_max_tx_power()`

Feel free to add it! It makes sense that Radio would have a tx_power attribute.

**Comment by askpatrickw** [comment]
I opened a Feature request for wifi.radio.tx_power() #6540

**Comment by askpatrickw** [comment]
OK, if this build works, and it should, I just tested wifi.radio.connect with the wifi.radio.tx_power, this is ready for final review.

**Review by tannewt — APPROVED** [review]
Looks good!

**Pull request merged**